gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java

```diff
@@ -0,0 +1,275 @@
+import gobblin.publisher.DataPublisher;
+import gobblin.publisher.UnpublishedHandling;
+import gobblin.runtime.commit.DatasetStateCommitStep;
+import gobblin.source.extractor.JobCommitPolicy;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * {@link Callable} that commits a single dataset. The logic in this class is thread-safe, however, it calls
+ * {@link DataPublisher#publish(Collection)}. This class is thread-safe if and only if the implementation of
+ * {@link DataPublisher} used is also thread-safe.
+ */
+@AllArgsConstructor
+@Slf4j
+final class SafeDatasetCommit implements Callable<Void> {
+
+  private final boolean shouldCommitDataInJob;
+  private final DeliverySemantics deliverySemantics;
+  private final String datasetUrn;
+  private final JobState.DatasetState datasetState;
+  private final boolean isMultithreaded;
+  private final JobContext jobContext;
+
+  @Override
+  public Void call()
+      throws Exception {
+
+    finalizeDatasetStateBeforeCommit(this.datasetState);
+
+    Class<? extends DataPublisher> dataPublisherClass;
+    try (Closer closer = Closer.create()) {
+      dataPublisherClass = JobContext.getJobDataPublisherClass(this.jobContext.getJobState())
+          .or((Class<? extends DataPublisher>) Class.forName(ConfigurationKeys.DEFAULT_DATA_PUBLISHER_TYPE));
+      if (!canCommitDataset(datasetState)) {
+        log.warn(String.format("Not committing dataset %s of job %s with commit policy %s and state %s",
+            this.datasetUrn, this.jobContext.getJobId(), this.jobContext.getJobCommitPolicy(), this.datasetState.getState()));
+        checkForUnpublishedWUHandling(this.datasetUrn, this.datasetState, dataPublisherClass, closer);
+        throw new RuntimeException(String.format("Not committing dataset %s of job %s with commit policy %s and state %s",
+            this.datasetUrn, this.jobContext.getJobId(), this.jobContext.getJobCommitPolicy(), this.datasetState.getState()));
+      }
+    } catch (ReflectiveOperationException roe) {
+      log.error("Failed to instantiate data publisher for dataset %s of job %s.", this.datasetUrn,
+          this.jobContext.getJobId(), roe);
+      throw new RuntimeException(roe);
+    }
+
+    Optional<CommitSequence.Builder> commitSequenceBuilder = Optional.absent();
+    try (Closer closer = Closer.create()) {
+      if (this.shouldCommitDataInJob) {
+        log.info(String.format("Committing dataset %s of job %s with commit policy %s and state %s", this.datasetUrn,
+            this.jobContext.getJobId(), this.jobContext.getJobCommitPolicy(), this.datasetState.getState()));
+        if (this.deliverySemantics == DeliverySemantics.EXACTLY_ONCE) {
+          generateCommitSequenceBuilder(this.datasetState);
+        } else {
+          DataPublisher publisher = closer.register(DataPublisher.getInstance(dataPublisherClass, this.jobContext.getJobState()));
+          if (this.isMultithreaded && !publisher.isThreadSafe()) {
```

Comment:
Why not move this check to `JobContext`, which is aware of publisher type as well? You can decide the real number of threads in `JobContext`, rather than adding `synchronized` committing here.